### PR TITLE
feat(provider/kubernetes): v2 restrict namespaces for uploading manifests

### PR DESCRIPTION
**Pull request by gardleopard**
Solves https://github.com/spinnaker/spinnaker/issues/2325

Depends on #2409

### CODE REVIEW

**Comment by gardleopard** [comment]
I have tested this now and it works, except I think the default namespace handling is strange.
It gets to this line https://github.com/spinnaker/clouddriver/blob/0a288b7e21d7746cc91eac6f3dce27afca840dd4/clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2Credentials.java#L127

I think maybe I cannot use that method and also that it needs to check for the ommitnamespaces and ignore all of this if no namespaces are explicitly defined.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java [hunk 1]
```diff
@@ -93,6 +100,11 @@ public boolean validateV2Credentials(AccountCredentialsProvider provider, String
       return false;
     }
 
+    if (!((KubernetesV2Credentials)credentials.getCredentials()).containsNamespace(namespace)) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: missing space after first `)`

**Comment by lwander** [comment]
Checking the omit namespace is definitely important - I don't think this should be the place to check the default namespace. I'm thinking some sort of validation should fail if someone specifies a "Default namespace" that they can't deploy to because it's not listed in the "namespaces", or omitted by the "omitNamespaces".

**Comment by gardleopard** [comment]
@lwander I have changed the logic and added tests for this now. I think the tests clearly shows the behaviour now.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2Credentials.java [hunk 2]
```diff
@@ -115,6 +115,14 @@ public String getDefaultNamespace() {
     return cachedDefaultNamespace;
   }
 
+  public List<String> getNamespaces() {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2Credentials.java]: I would just annotated the `namespaces` and `omitNamespaces` fields with `@Getter` like the properties above around line 77

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java [hunk 3]
```diff
@@ -93,6 +100,25 @@ public boolean validateV2Credentials(AccountCredentialsProvider provider, String
...
+      return false;
+    }
+
+    return true;
+  }
+
+  protected boolean validateNamespace(String namespace, KubernetesV2Credentials credentials) {
+    final List<String> configuredNamespaces = credentials.getNamespaces();
+    if (configuredNamespaces != null && !configuredNamespaces.isEmpty() && !configuredNamespaces.contains(namespace)) {
+      reject( "wrongNamespace", "namespace");
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: Extra space before `"`

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java [hunk 4]
```diff
@@ -93,6 +100,25 @@ public boolean validateV2Credentials(AccountCredentialsProvider provider, String
...
+      return false;
+    }
+
+    final List<String> omitNamespaces = credentials.getOmitNamespaces();
+    if (omitNamespaces != null && omitNamespaces.contains(namespace)) {
+      reject( "omittedNamespace", "namespace");
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: Extra space before `"`

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java [hunk 5]
```diff
@@ -77,11 +79,16 @@ public boolean validateNotEmpty(String attribute, String value) {
...
   }
 
-  public boolean validateV2Credentials(AccountCredentialsProvider provider, String accountName) {
+  public boolean validateV2Credentials(AccountCredentialsProvider provider, String accountName, String namespace) {
+    log.info("Validating credentials for {} {}", accountName, namespace);
     if (!validateNotEmpty("account", accountName)) {
       return false;
     }
 
+    if (!validateNotEmpty("account", namespace)) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: Should this be `"namespace"`?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java (hunk 3 above)
Comment (lwander) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: Also, should the string `"namespace`" be replaced with the actual requested namespace?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java [hunk 6]
```diff
@@ -93,6 +100,25 @@ public boolean validateV2Credentials(AccountCredentialsProvider provider, String
...
     }
 
+    if (!validateNamespace(namespace, (KubernetesV2Credentials)credentials.getCredentials())) {
+      return false;
+    }
+
+    return true;
+  }
+
+  protected boolean validateNamespace(String namespace, KubernetesV2Credentials credentials) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: I'm thinking we should skip this (e.g. `return true`) when `namespace` is empty. If none is supplied, the default will be used which should be valid.

Comment (gardleopard) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: in line 88 we check if the namespace is empty, so it wont reach this part of the code if a namespace is not supplied.

Comment (lwander) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: Ah I missed that - thanks!

Comment (wjoel) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/validator/KubernetesValidationUtil.java]: Line 89 returns `false`, which is not correct for resources without a namespace. I've submitted https://github.com/spinnaker/clouddriver/pull/2422 to change this.

**Pull request merged**